Atlas/android/app/src/main/java/com/bounswe2017/group10/atlas/response/OnCreateItemResponse.java

```diff
@@ -48,41 +47,27 @@ public void onResponse(Call<CreateItemResponse> call, Response<CreateItemRespons
             } else {
                 int id = response.body().getId();
 
-                RandomStringGenerator generator = new RandomStringGenerator.Builder()
-                        .withinRange('a', 'z')
-                        .build();
-                List<Image> localImages = new ArrayList<>();
-                List<String> filenames = new ArrayList<>();
-                for (Image img : this.mImageList) {
-                    if (Utils.isLocalUrl(img.getUrl())) {
-                        filenames.add(img.getUrl());
-
-                        String filename = generator.generate(Constants.CLOUDINARY_IMG_NAME_LENGTH);
-                        img.setUrl(Utils.filenameToCloudinaryUrl(filename));
-                        localImages.add(img);
+                int lastLocalUrl = 0;
+                for (int i = 0; i < mImageList.size(); ++i) {
+                    if (Utils.isLocalUrl(mImageList.get(i).getUrl())) {
+                        lastLocalUrl = i;
+                        break;
                     }
                 }
-
-                // share an atomic counter between response handlers
-                AtomicInteger uploadCount = new AtomicInteger(0);
                 OnCloudinaryUploadResponse respHandler = new OnCloudinaryUploadResponse(
                         context,
                         createFragment,
                         mImageList,
                         progressBar,
                         id,
-                        uploadCount,
-                        localImages.size()
+                        lastLocalUrl
                 );
-                // dispatch each request after creating a resp handler for each of them.
-                for (int i = 0; i < localImages.size(); ++i) {
-                    MediaManager.get()
-                            .upload(Uri.parse(filenames.get(i)))
-                            .unsigned("wak3gala")
-                            .option("public_id", Utils.cloudinaryUrlToFilename(localImages.get(i).getUrl()))
-                            .callback(respHandler)
-                            .dispatch();
-                }
+                String fileUrl = mImageList.get(lastLocalUrl).getUrl();
+                MediaManager.get()
```

Comment:
For now it is not a high priority but if we are uploading a lot of images it will certainly take a long time, so why not do uploading in new threads to make things faster ? I think they are not dependent and it looks more convenient to me.

Reply:
This got outdated, but take a look at this.